Implemented Jaccard distance

lvoursl opened the pull request:
**Summary**
Implemented Jaccard distance requested in this issue: https://github.com/facebookresearch/faiss/issues/1299
**Test plan**
Run: make -C build test
Output: 100% tests passed, 0 tests failed out of 174

mdouze reviewed (commented):
please add the inline function as discussed in the chat.

mdouze commented on faiss/gpu/impl/DistanceUtils.cuh:
```diff
@@ -251,6 +251,35 @@ struct JensenShannonDistance {
     float dist;
 };
 
+struct JaccardDistance {
```
Let's stick with "Jaccard", since that's how the issue refers to it, and comment what it is in MetricType.h

lvoursl commented:
I've changed the names, added more comments and function which we've discussed in a chat :) 

Do you want to replace all checks for inner product with `is_dissimilairty_metric` function in this PR or shall we have separate one?

mdouze commented:
I'm fine with having all checks in this PR.

lvoursl commented:
I've replaced checks for inner product with `is_dissimilairty_metric` where it's generic case and there is no inner product specific behaviour later on. It would be great to have second eyes look on it since I'm quite new to the codebase.

mdouze commented on faiss/MetricType.h:
```diff
@@ -31,11 +31,19 @@ enum MetricType {
+/// this function is used to distinguish between min and max indexes since
+/// we need to support similarity and dis-similarity metrics in a flexible way
+constexpr bool is_dissimilairty_metric(MetricType metric_type) {
+    return metric_type == METRIC_INNER_PRODUCT;
```
Hmm it should be `is_similarity_metric` in fact, since larger values of the metric are better...
Also, Jaccard is also a similarity metric no?

mdouze reviewed (commented):
Quite a few CI tests are failing, they seem to be related to the HNSW and NSG implementations.

mdouze commented:
Looks good now. Please import the code to fbcode then we can land it there.

lvoursl commented:
Sounds good, will do!
